### PR TITLE
Refactor indexing

### PR DESCRIPTION
**Pull request by tkporter**
### Description

* Stops creating a new client a bunch of times in rpc.rs (I haven't yet changed grpc.rs - this is because the client's functions are `&mut self`, so will have to think about what to do here)
* Changes `unwrap_or_none_result` to not include the `let Some(varname)` bit so we can do `let varname = unwrap_or_none_result!(...)`
* Instead of a `get_parser(&self, ...)` fn that returns a closure, we just pass a the function `Self::[data_type]_parser`
* Refactors these parsers to be a bit more defensive - e.g. each param is represented as an Option so it's clear if one isn't parsed, and we also get the `_contract_address` param (which is auto-set by the cosmwasm runtime). It returns a `ParsedEvent<T>` so that we can return the `_contract_address` in there too
* Refactors the range indexing in rpc.rs, including an important fix to ensure that the contract address of an event is the one we actually want to index

### Drive-by changes

n/a

### Related issues

n/a

### Backward compatibility

Fully

### Testing

Unit tests, ran locally

### CODE REVIEW

File: rust/chains/hyperlane-cosmos/src/providers/rpc.rs [hunk 1]
```diff
@@ -61,41 +67,101 @@ impl CosmosWasmIndexer {
...
+                // Filter out failed txs
+                if tx.tx_result.code.is_err() {
+                    debug!(tx_hash=?tx.hash, "Indexed tx has failed, skipping");
+                    false
+                } else {
+                    true
+                }
+            })
+            .map(move |tx| {
+                // Iter through all events in the tx, looking for the target event.
```
Comment (daniel-savu) [inline rust/chains/hyperlane-cosmos/src/providers/rpc.rs:116]: Could remove this comment if the closure is moved to its own function named appropriately, and we'd also get less scary nesting

**Review by daniel-savu — APPROVED** [review]

File: rust/chains/hyperlane-cosmos/src/interchain_gas.rs [hunk 2]
```diff
@@ -51,6 +56,24 @@ impl CosmosInterchainGasPaymaster {
     }
 }
 
+// ------------------ Indexer ------------------
+
+const MESSAGE_ID_ATTRIBUTE_KEY: &str = "message_id";
+// echo -n message_id | base64
+const MESSAGE_ID_ATTRIBUTE_KEY_BASE64: &str = "bWVzc2FnZV9pZA==";
```
[the remark this answers is not in the document]
Comment (tkporter) [reply rust/chains/hyperlane-cosmos/src/interchain_gas.rs:63]: Ah good idea, i'll move to this

File: rust/chains/hyperlane-cosmos/src/providers/rpc.rs [hunk 3]
```diff
@@ -107,93 +173,47 @@ impl WasmIndexer for CosmosWasmIndexer {
...
         }
 
-        Ok(result)
+        Ok(logs)
     }
 }
+
+fn div_ceil(numerator: u32, denominator: u32) -> u32 {
+    (numerator + denominator - 1) / denominator
+}
```
[the remark this answers is not in the document]
Comment (tkporter) [reply rust/chains/hyperlane-cosmos/src/providers/rpc.rs:219]: Imo not dealing with floats is far nicer 🙈

[the remark this answers is not in the document]
Comment (tkporter) [reply rust/chains/hyperlane-cosmos/src/providers/rpc.rs:219]: Relying on floats and casting to u32 feels sketchier

File: rust/chains/hyperlane-cosmos/src/interchain_gas.rs [hunk 4]
```diff
@@ -64,56 +87,104 @@ impl CosmosInterchainGasPaymasterIndexer {
...
+                DESTINATION_ATTRIBUTE_KEY_BASE64 => {
+                    destination = Some(String::from_utf8(BASE64.decode(value)?)?.parse()?);
+                }
+
+                _ => {}
+            }
         }
+
+        let contract_address = unwrap_or_none_result!(
+            contract_address,
```
[the remark this answers is not in the document]
Comment (tkporter) [reply rust/chains/hyperlane-cosmos/src/interchain_gas.rs:158]: mm good point this feels cleaner, will do this

File: rust/chains/hyperlane-cosmos/src/providers/rpc.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniel-savu) [reply rust/chains/hyperlane-cosmos/src/providers/rpc.rs:219]: `ceil()` will return an integer though

**Pull request merged**